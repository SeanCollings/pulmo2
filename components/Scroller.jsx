import React, { useState, useEffect, useRef } from 'react';
import { StyleSheet, ScrollView, View } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { TouchableOpacity } from 'react-native';

import { COLORS } from '../constants/constants';

const Scroller = ({
  children,
  totalChildren,
  onScroll,
  initialIndex,
  width = 80,
}) => {
  const infListRef = useRef();
  const [currentIndex, setCurrentIndex] = useState(initialIndex);

  useEffect(() => {
    infListRef.current.scrollTo({
      x: currentIndex * width,
      y: 0,
      animated: false,
    });
  }, []);

  const onLeftPressHandler = () => {
    if (currentIndex > 0) {
      const newIndex = currentIndex - 1;
      infListRef.current.scrollTo({
        x: newIndex * width,
        y: 0,
        animated: true,
      });
      setCurrentIndex(newIndex);
      onScroll(newIndex);
    }
  };
  const onRightPressHandler = () => {
    if (currentIndex < totalChildren - 1) {
      const newIndex = currentIndex + 1;
      infListRef.current.scrollTo({
        x: newIndex * width,
        y: 0,
        animated: true,
      });

      setCurrentIndex(newIndex);
      onScroll(newIndex);
    }
  };

  return (
    <View style={{ ...styles.container }}>
      <TouchableOpacity style={styles.leftButton} onPress={onLeftPressHandler}>
        <MaterialCommunityIcons
          name={'chevron-left-circle'}
          color={COLORS.BORDER}
          size={30}
        />
      </TouchableOpacity>
      <View style={{ width }}>
        <ScrollView
          pagingEnabled
          horizontal
          showsHorizontalScrollIndicator={false}
          scrollEnabled={false}
          ref={infListRef}
          keyboardShouldPersistTaps="always"
        >
          {children}
        </ScrollView>
      </View>
      <TouchableOpacity
        style={styles.rightButton}
        onPress={onRightPressHandler}
      >
        <MaterialCommunityIcons
          name={'chevron-right-circle'}
          color={COLORS.BORDER}
          size={30}
        />
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'flex-start',
  },
  scrollViewContainer: {},
  leftButton: {
    justifyContent: 'center',
    paddingTop: 4,
  },
  rightButton: {
    justifyContent: 'center',
    paddingTop: 4,
  },
  childrenContainer: {
    padding: 10,
    backgroundColor: 'orange',
  },
});

export default Scroller;
