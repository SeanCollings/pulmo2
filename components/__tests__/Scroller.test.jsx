import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup, fireEvent } from 'react-native-testing-library';
import Scroller from '../Scroller';
import { Text } from 'react-native';

describe('Scroller - unit test', () => {
  let component;

  const children = [
    <Text key={0}>one</Text>,
    <Text key={1}>two</Text>,
    <Text key={2}>three</Text>,
  ];

  afterEach(() => {
    cleanup();
  });

  test('should render', async () => {
    const onScroll = jest.fn();

    await act(async () => {
      component = create(
        <Scroller
          children={children}
          totalChildren={children.length}
          onScroll={onScroll}
          initialIndex={0}
        />
      );
    });
    expect(component).toMatchSnapshot();
  });

  test('should handle left press if index greater than 0', async () => {
    const onScroll = jest.fn();

    await act(async () => {
      component = create(
        <Scroller
          children={children}
          totalChildren={children.length}
          onScroll={onScroll}
          initialIndex={1}
        />
      );
    });

    const leftButton = component.root.find(
      ({ props }) => props.testID === 'left'
    );

    fireEvent.press(leftButton);
    expect(onScroll).toHaveBeenCalled();
  });

  test('should not call onScroll on left press if index equal to 0', async () => {
    const onScroll = jest.fn();

    await act(async () => {
      component = create(
        <Scroller
          children={children}
          totalChildren={children.length}
          onScroll={onScroll}
          initialIndex={0}
        />
      );
    });

    const leftButton = component.root.find(
      ({ props }) => props.testID === 'left'
    );

    fireEvent.press(leftButton);
    expect(onScroll).toHaveBeenCalledTimes(0);
  });

  test('should handle right press', async () => {
    const onScroll = jest.fn();

    await act(async () => {
      component = create(
        <Scroller
          children={children}
          totalChildren={children.length}
          onScroll={onScroll}
          initialIndex={1}
        />
      );
    });

    const rightButton = component.root.find(
      ({ props }) => props.testID === 'right'
    );

    fireEvent.press(rightButton);
    expect(onScroll).toHaveBeenCalled();
  });

  test('should not call onScroll on right press if index equal to children length', async () => {
    const onScroll = jest.fn();

    await act(async () => {
      component = create(
        <Scroller
          children={children}
          totalChildren={children.length}
          onScroll={onScroll}
          initialIndex={children.length - 1}
        />
      );
    });

    const rightButton = component.root.find(
      ({ props }) => props.testID === 'right'
    );

    fireEvent.press(rightButton);
    expect(onScroll).toHaveBeenCalledTimes(0);
  });
});
