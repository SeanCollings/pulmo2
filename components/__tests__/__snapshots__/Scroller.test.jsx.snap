// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroller - unit test should render 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "flexDirection": "row",
      "justifyContent": "flex-start",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "justifyContent": "center",
        "opacity": 1,
        "paddingTop": 4,
      }
    }
    testID="left"
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#CCC",
            "fontSize": 30,
          },
          undefined,
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ﬍
    </Text>
  </View>
  <View
    style={
      Object {
        "width": 80,
      }
    }
  >
    <RCTScrollView
      horizontal={true}
      keyboardShouldPersistTaps="always"
      pagingEnabled={true}
      scrollEnabled={false}
      showsHorizontalScrollIndicator={false}
    >
      <View>
        <Text>
          one
        </Text>
        <Text>
          two
        </Text>
        <Text>
          three
        </Text>
      </View>
    </RCTScrollView>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "justifyContent": "center",
        "opacity": 1,
        "paddingTop": 4,
      }
    }
    testID="right"
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#CCC",
            "fontSize": 30,
          },
          undefined,
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ﬏
    </Text>
  </View>
</View>
`;
